[quant][core][feature] Implement index_put for quantized CUDA tensors

jcaip opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85164

Summary:
- Add new cuda test for quantized index_put
- Add determinsitc test for CPU and CUDA quantized index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`
    - CUDA index_put determinstic implemented in `index_put_with_sort_kernel_quantized`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits. Might be something useful to add in the
future?

Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put
```

Reviewers:

Subscribers:

Tasks:

Tags: quant

jcaip committed:
[quant][core][feature] Implement index_put for quantized CUDA tensors

Summary:
- Add new cuda test for index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits. Might be something useful to add in the
future?

Test Plan:
``
python test/test_quantization.py -k test_qtensor_index_put

``

Reviewers:

Subscribers:

Tasks:

Tags: quant`

[ghstack-poisoned]

jcaip committed:
Update on "[quant][core][feature] Implement index_put for quantized CUDA tensors"


Summary:
- Add new cuda test for index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits, so I had to reimplement some of the quantization code. 


Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put

```

Reviewers:

Subscribers:

Tasks:

Tags: quant`

[ghstack-poisoned]

jcaip committed:
Update on "[quant][core][feature] Implement index_put for quantized CUDA tensors"


Summary:
- Add new cuda test for index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits, so I had to reimplement some of the quantization code. 


Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put

```

Reviewers:

Subscribers:

Tasks:

Tags: quant`

[ghstack-poisoned]

jcaip committed:
Update on "[quant][core][feature] Implement index_put for quantized CUDA tensors"


Summary:
- Add new cuda test for index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits, so I had to reimplement some of the quantization code. 


Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put

```

Reviewers:

Subscribers:

Tasks:

Tags: quant`

[ghstack-poisoned]

jcaip committed:
Update on "[quant][core][feature] Implement index_put for quantized CUDA tensors"


Summary:
- Add new cuda test for index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits, so I had to reimplement some of the quantization code. 


Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put

```

Reviewers:

Subscribers:

Tasks:

Tags: quant`

[ghstack-poisoned]

jcaip committed:
Update on "[quant][core][feature] Implement index_put for quantized CUDA tensors"


Summary:
- Add new cuda test for quantized index_put
- Add determinsitc test for CPU and CUDA quantized index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`
    - CUDA index_put determinstic implemented in `index_put_with_sort_kernel_quantized`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits. Might be something useful to add in the
future?

Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put
```

Reviewers:

Subscribers:

Tasks:

Tags: quant

[ghstack-poisoned]

jcaip committed:
Update on "[quant][core][feature] Implement index_put for quantized CUDA tensors"


Summary:
- Add new cuda test for quantized index_put
- Add determinsitc test for CPU and CUDA quantized index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`
    - CUDA index_put determinstic implemented in `index_put_with_sort_kernel_quantized`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits. Might be something useful to add in the
future?

Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put
```

Reviewers:

Subscribers:

Tasks:

Tags: quant

[ghstack-poisoned]

jcaip committed:
Update on "[quant][core][feature] Implement index_put for quantized CUDA tensors"


Summary:
- Add new cuda test for quantized index_put
- Add determinsitc test for CPU and CUDA quantized index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`
    - CUDA index_put determinstic implemented in `index_put_with_sort_kernel_quantized`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits. Might be something useful to add in the
future?

Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put
```

Reviewers:

Subscribers:

Tasks:

Tags: quant

[ghstack-poisoned]

dzdang commented on aten/src/ATen/native/TensorAdvancedIndexing.cpp:
```diff
@@ -416,6 +416,7 @@ DEFINE_DISPATCH(put_stub);
 DEFINE_DISPATCH(take_stub);
 DEFINE_DISPATCH(masked_fill_stub);
 REGISTER_NO_CPU_DISPATCH(index_put_with_sort_stub);
+REGISTER_NO_CPU_DISPATCH(index_put_with_sort_kernel_quantized_stub);
```
nit: remove kernel in name to match L418?

dzdang commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -239,6 +240,21 @@ static void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntAr
+    float inv_scale = 1.0f / static_cast<float>(scale);
+
+    gpu_index_kernel(iter, index_size, index_stride, [inv_scale, zero_point, qmin, qmax]C10_DEVICE(char* out_data, char* in_data, int64_t offset) {
+      int64_t qvalue = static_cast<int64_t>(zero_point + nearbyintf(*(float*)in_data * inv_scale));
```
is the `nearbyintf` necessary since there's already a static_cast to int64_t?

dzdang commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -239,6 +240,21 @@ static void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntAr
+
+    gpu_index_kernel(iter, index_size, index_stride, [inv_scale, zero_point, qmin, qmax]C10_DEVICE(char* out_data, char* in_data, int64_t offset) {
+      int64_t qvalue = static_cast<int64_t>(zero_point + nearbyintf(*(float*)in_data * inv_scale));
+      qvalue = min(max(qvalue, qmin), qmax);
```
think we can just use `at::clamp`

jcaip replied:
I get an error: `namespace "at" has no member "clamp"`

dzdang commented on aten/src/ATen/native/cuda/IndexKernel.cu:
```diff
@@ -239,6 +240,21 @@ static void index_put_kernel(TensorIterator& iter, IntArrayRef index_size, IntAr
+    gpu_index_kernel(iter, index_size, index_stride, [inv_scale, zero_point, qmin, qmax]C10_DEVICE(char* out_data, char* in_data, int64_t offset) {
+      int64_t qvalue = static_cast<int64_t>(zero_point + nearbyintf(*(float*)in_data * inv_scale));
+      qvalue = min(max(qvalue, qmin), qmax);
+      *(scalar_t*)(out_data + offset) = static_cast<scalar_t>(qvalue);
```
can we use ` *static_cast<scalar_t*>(out_data + offset)`. a little clearer to me

jcaip replied:
Also get a compilation error when I do this, it complains about an invalid type conversion.

dzdang replied:
can you paste the error here?

dzdang replied:
I see. I think we can just keep it the way it is. looks like there's some lint errors, but other than that should be good to land

dzdang reviewed (approved):
look great! had a few minor comments

jcaip committed:
Update on "[quant][core][feature] Implement index_put for quantized CUDA tensors"

Summary:
- Add new cuda test for quantized index_put
- Add determinsitc test for CPU and CUDA quantized index_put
- Add in QuantizedCUDA implementation for index_put
    - wrote new `index_put_kernel_quantized_cuda`
    - CUDA index_put determinstic implemented in `index_put_with_sort_kernel_quantized`

I think quantize_val<scalar_t> is not CUDA compatible, because of the
reliance on std::numeric_limits. Might be something useful to add in the
future?

Test Plan:
```
python test/test_quantization.py -k test_qtensor_index_put
```

Reviewers:

Subscribers:

Tasks:

Tags: quant

[ghstack-poisoned]

jcaip commented:
@pytorchbot merge

jcaip commented:
This was merged in https://github.com/pytorch/pytorch/pull/85685
